File: 238_product_of_array_except_self.py
from typing import List

def product_except_self(nums: List[int]) -> List[int]:
    le = len(nums)

    pre = [1 * nums[0]]
    for i in range(1, le - 1):
        pre.append(nums[i] * pre[i - 1])

    post = [1 for i in range(le)]
    post[le - 1] = nums[le - 1]
    for i in range(le - 2, -1, -1):
        post[i] = nums[i] * post[i + 1]

    res = [1 for i in range(le)]
    res[0] = post[1]
    res[le - 1] = pre[le - 2]
    for i in range(1, le - 1):
        res[i] = pre[i - 1] * post[i + 1]

    return res


print(product_except_self([-1,1,0,-3,3]))

File: go/238_product_of_array_except_self.go
package main

func productExceptSelf(nums []int) []int {
	n := len(nums)
	res := make([]int, n)

	asc := make([]int, n)
	asc[0] = nums[0]
	for i := 1; i < n; i++ {
		asc[i] = asc[i-1] * nums[i]
	}

	desc := make([]int, n)
	desc[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		desc[i] = desc[i+1] * nums[i]
	}

	res[0] = desc[1]
	res[n-1] = asc[n-2]
	for i := 1; i < n-1; i++ {
		res[i] = asc[i-1] * desc[i+1]
	}

	return res
}
